text/0000-unsafe-attributes.md

```diff
@@ -0,0 +1,198 @@
+`unsafe(attr)` anywhere that `attr` can be used. `unsafe` cannot be nested,
+cannot contain `cfg_attr`, and cannot contain any other (non-unsafe) attributes.
+
+The `deny(unsafe_code)` lint denies the use of unsafe attributes both inside and
+outside of `unsafe(...)` blocks. (That lint currently has special handling to
+deny these attributes. Once there is a general notion of 'unsafe attributes' as
+proposed by this RFC, that special handling should no longer be needed.)
+
+The `unsafe(...)` attribute block is required even for functions declared inside
+an `unsafe` block. That is, the following is an error:
+
+```rust
+fn outer() {
+  unsafe {
+    #[no_mangle]
+    fn write() {}
+  }
+}
+```
+
+This matches the fact that expression-level unsafety is not inherited for items
+declared inside other items.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+I think if we had thought of this around Rust 1.0, then this would be rather
+uncontroversial. As things stand now, this proposal will cause a lot of churn
+since all existing uses of these unsafe attributes need to be adjusted. The
+warning for using unsafe attributes outside `unsafe(...)` should probably have
+an auto-fix available to help ease the transition here.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- **Nothing.** We could do nothing at all, and live with the status quo. However
+  then we will not be able to fix issues like
+  [`no_mangle` being unsound](https://github.com/rust-lang/rust/issues/28179),
+  which is one of the oldest open soundness issues.
+- **Rename.** We could just rename the attributes to `unsafe_no_mangle` etc.
+  However that is inconsistent with how we approach `unsafe` on expressions, and
+  feels much less systematic and much more ad-hoc.
+- **`deny(unsafe_code)`.** We already
+  [started the process](https://github.com/rust-lang/rust/issues/82499) of
+  rejecting these attributes when `deny(unsafe_code)` is used. We could say that
+  is enough. However the RFC authors thinks that is insufficient, since only few
+  crates use that lint, and since it is the wrong default for Rust (users have
+  to opt-in to a soundness-critical diagnostic -- that's totally against the
+  "safety by default" goal of Rust). This RFC says that yes, `deny(unsafe_code)`
+  should deny those attributes, but we should go further and require an explicit
+  `unsafe(...)` attribute block for them to be used at all.
+- **Item-level unsafe blocks.** We could find some way to have 'unsafe blocks'
+  around entire functions or modules. However, those would go against the usual
+  goal of keeping `unsafe` blocks small. Big `unsafe` blocks risk accidentally
+  calling an unsafe operation in there without even realizing it.
+- **Other syntax.** Obviously we could pick a different syntax for the same
+  concept, but this seems like the most natural marriage of the idea of unsafe
+  blocks from regular code, and the existing attributes syntax.
+
```

Comment:
Pondering: this leans more into the use of `unsafe` as the keyword to discharge the safety obligation.

Would this be a place to add the `hold_my_beer` (not under that name) term that is for discarding specifically, to leave `unsafe` as the marker for introducing obligations?

Or would it be too weird to have the different word *here* but not in the other places?

Reply:
> Or would it be too weird to have the different word here but not in the other places?

Yeah that feels pretty weird to me. If we want to use another keyword for this use of unsafe we should do it consistently.